Added HealthCommand, based off the existing FreezeCommand for modifyi…

Melodi17 opened the pull request:
…ng living entities' health.

This is a useful command in my opinion as it adds an easy way to modify any entities' health even players.

Commit:
Added HealthCommand, based off the existing FreezeCommand for modifying living entities' health.
This is a useful command in my opinion as it adds an easy way to modify any entities' health even players.

Melodi17 committed:
Merge remote-tracking branch 'upstream/master'

TheDeathlyCow commented:
Thanks for this! I will try and get to it later today.

Melodi17 commented:
Sorry, wrong button

Melodi17 committed:
Patched chat message '.. (now %d)'  health command output

Previously the output message displayed the max health of the entity instead of the current health.

TheDeathlyCow commented on src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java:
```diff
@@ -0,0 +1,281 @@
+    private static int set(final ServerCommandSource source, final Collection<? extends Entity> targets, final int amount, final boolean shouldClamp) {
+        int sum = 0;
+        for (Entity target : targets) {
+            if (!(target instanceof LivingEntity livingTarget)) {
```
Either all `LivingEntity`s selected by this command should be affected by its execution, or no entities should be affected. It should not return early. Also, the message should only show if a single entity is selected.

Melodi17 replied:
Fixed, the code now checks the targets beforehand

TheDeathlyCow commented on src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java:
```diff
@@ -0,0 +1,281 @@
+    private static int adjust(final ServerCommandSource source, final Collection<? extends Entity> targets, final int amount, final boolean shouldClamp, final boolean isRemoving) {
+        int sum = 0;
+        for (Entity target : targets) {
+            if (!(target instanceof LivingEntity livingTarget)) {
```
Same as previous comment.

Melodi17 replied:
Fixed, same as previous comment

TheDeathlyCow commented on src/main/java/com/github/thedeathlycow/tdcdata/server/command/HealthCommand.java:
```diff
@@ -0,0 +1,281 @@
+            return -1;
+        }
+
+        int amount = Math.round(((LivingEntity) target).getHealth());
```
I'm not sure `Math.round` is the best function to use here; perhaps `MathHelper.floor` should be used instead? Reason being that if you have 19.5 health, your health bar will render 19 health points. However, `Math.round` would round 19.5 to 20.

Melodi17 replied:
Addressed this, changed `Math.round` to `MathHelper.floor`

Melodi17 commented:
[Health-Command.md](https://github.com/TheDeathlyCow/tdc-data/files/9075350/Health-Command.md)
Here is a wiki page I wrote for it (based on the freeze page), if that helps.

Melodi17 committed:
Merge remote-tracking branch 'upstream/master'

Melodi17 committed:
Addressed @TheDeathlyCow feedback

Melodi17 committed:
Patched health command output message for single entity selection

TheDeathlyCow commented:
I've been thinking some more about how this command could handle living vs non living entities, and I think perhaps the best thing to do might be to make the target of the command a single entity for all cases. The vanilla command [/attribute](https://minecraft.fandom.com/wiki/Commands/attribute) is a somewhat similar idea to this, and it only allows a single entity to be targeted at a time. You can of course use `/execute as` to target multiple entities, and any non living targets will effectively be skipped. It is a little more laggy than targeting multiple entities, but still much better than `/data` and in my opinion more consistent with vanilla.

Melodi17 commented:
I like this idea, error handling/messages will be less complex and it will seem to be a more "vanilla" feature. For consistency, should we do something similar with the `/freeze` command? I will start implementing your idea as soon as I can (5-9 hours from this message).

Melodi17 committed:
Implemented @TheDeathlyCow s feedback

- Set and Adjust methods take a singular entity at a time, like the /data command for consistency, makes the code cleaner and more clear functionality

- Adjusted the return value of the Get functions to return 0 instead of -1 in the event of an error occuring, for consistency with the other functions.

Melodi17 committed:
Implemented @TheDeathlyCow s feedback

- Set and Adjust methods take a singular entity at a time, like the /attribute command for consistency, makes the code cleaner and more clear functionality

- Adjusted the return value of the Get functions to return 0 instead of -1 in the event of an error occuring, for consistency with the other functions.

Melodi17 committed:
Merge branch 'master' of https://github.com/Melodi17/tdc-data

TheDeathlyCow commented:
The reason `/freeze` can still target any entity is because, unlike health, freezing is something any `Entity` can do, it is not limited to a certain subclass of `Entity`.

TheDeathlyCow committed:
Use command syntax exceptions on fail

Instead of displaying an error to the player when the target of /health is not a living entity, throw an exception with the message. Minecraft will handle the rest.

TheDeathlyCow committed:
fixed variable names

oops
